subsys/power/power.c

```diff
@@ -116,6 +120,18 @@ enum power_states _sys_suspend(s32_t ticks)
 		 * in deep sleep mode.
 		 */
 		_sys_pm_idle_exit_notification_disable();
+#if CONFIG_DEVICE_POWER_MANAGEMENT
+	} else {
+		if (sys_pm_policy_low_power_devices(pm_state)) {
+			/* low power peripherals. */
+			if (sys_pm_low_power_devices()) {
+				LOG_DBG("Device enter low power state failed!");
+				sys_pm_resume_devices();
```

Comment:
@vanti sys_pm_suspend_devices will be in the environment of power gating most power domain, so devices can't work if system is in that state. But sys_pm_low_power_devices will be in state of clock gate or reduce clock, power is on, devices can work in that case, so even fail to low power devices, we still can have system enter sleep state as the same behavior before this change.

Reply:
I don't think it is strictly true that devices can work for any sleep state, at least per my understanding thus far. For instance, on the TI CC1352, there is a sleep state where it can enter a standby mode in which the CPU does not entirely lose its context (ie. it can resume from the point where it went to sleep), but some devices, such as UART/I2C/SPI do not work.  

That said, I expect this code would still work for CC1352, because its implementation of 'set_power_state' does not return failure. But I wonder if we are making an assumption that would cause an error to be masked on other platforms.

Reply:
what I thought is when CPU core only is in clock gate state, some devices can do hardware transfer (such as DMA can still do data transfer because before enter sleep state, someone gives DMA descriptor), when done the transfer, interrupt will come and wake-up CPU. If there is no that case, I will change the code. @vanti

Reply:
@vanti Maybe I should consider more, just be consistent with deep sleep currently, updated code now.